// humanloop/serialization/types/EvaluationRunResponse.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.EvaluationRunResponse = void 0;
const core = __importStar(require("../../core"));
const DatasetResponse_1 = require("./DatasetResponse");
const RunVersionResponse_1 = require("./RunVersionResponse");
const UserResponse_1 = require("./UserResponse");
const EvaluationStatus_1 = require("./EvaluationStatus");
exports.EvaluationRunResponse = core.serialization.object({
    id: core.serialization.string(),
    dataset: DatasetResponse_1.DatasetResponse.optional(),
    version: RunVersionResponse_1.RunVersionResponse.optional(),
    orchestrated: core.serialization.boolean(),
    addedAt: core.serialization.property("added_at", core.serialization.date()),
    createdAt: core.serialization.property("created_at", core.serialization.date()),
    createdBy: core.serialization.property("created_by", UserResponse_1.UserResponse.optional()),
    status: EvaluationStatus_1.EvaluationStatus,
    control: core.serialization.boolean(),
});
