// humanloop/serialization/types/HumanEvaluatorRequest.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.HumanEvaluatorRequest = void 0;
const core = __importStar(require("../../core"));
const EvaluatorArgumentsType_1 = require("./EvaluatorArgumentsType");
const HumanEvaluatorRequestReturnType_1 = require("./HumanEvaluatorRequestReturnType");
const EvaluatorJudgmentOptionResponse_1 = require("./EvaluatorJudgmentOptionResponse");
const EvaluatorJudgmentNumberLimit_1 = require("./EvaluatorJudgmentNumberLimit");
const Valence_1 = require("./Valence");
exports.HumanEvaluatorRequest = core.serialization.object({
    argumentsType: core.serialization.property("arguments_type", EvaluatorArgumentsType_1.EvaluatorArgumentsType),
    returnType: core.serialization.property("return_type", HumanEvaluatorRequestReturnType_1.HumanEvaluatorRequestReturnType),
    attributes: core.serialization.record(core.serialization.string(), core.serialization.unknown()).optional(),
    options: core.serialization.list(EvaluatorJudgmentOptionResponse_1.EvaluatorJudgmentOptionResponse).optional(),
    numberLimits: core.serialization.property("number_limits", EvaluatorJudgmentNumberLimit_1.EvaluatorJudgmentNumberLimit.optional()),
    numberValence: core.serialization.property("number_valence", Valence_1.Valence.optional()),
    evaluatorType: core.serialization.property("evaluator_type", core.serialization.stringLiteral("human")),
    instructions: core.serialization.string().optional(),
});
