// humanloop/serialization/types/PromptKernelRequest.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.PromptKernelRequest = void 0;
const core = __importStar(require("../../core"));
const ModelEndpoints_1 = require("./ModelEndpoints");
const PromptKernelRequestTemplate_1 = require("./PromptKernelRequestTemplate");
const TemplateLanguage_1 = require("./TemplateLanguage");
const ModelProviders_1 = require("./ModelProviders");
const PromptKernelRequestStop_1 = require("./PromptKernelRequestStop");
const ResponseFormat_1 = require("./ResponseFormat");
const ReasoningEffort_1 = require("./ReasoningEffort");
const ToolFunction_1 = require("./ToolFunction");
exports.PromptKernelRequest = core.serialization.object({
    model: core.serialization.string(),
    endpoint: ModelEndpoints_1.ModelEndpoints.optional(),
    template: PromptKernelRequestTemplate_1.PromptKernelRequestTemplate.optional(),
    templateLanguage: core.serialization.property("template_language", TemplateLanguage_1.TemplateLanguage.optional()),
    provider: ModelProviders_1.ModelProviders.optional(),
    maxTokens: core.serialization.property("max_tokens", core.serialization.number().optional()),
    temperature: core.serialization.number().optional(),
    topP: core.serialization.property("top_p", core.serialization.number().optional()),
    stop: PromptKernelRequestStop_1.PromptKernelRequestStop.optional(),
    presencePenalty: core.serialization.property("presence_penalty", core.serialization.number().optional()),
    frequencyPenalty: core.serialization.property("frequency_penalty", core.serialization.number().optional()),
    other: core.serialization.record(core.serialization.string(), core.serialization.unknown()).optional(),
    seed: core.serialization.number().optional(),
    responseFormat: core.serialization.property("response_format", ResponseFormat_1.ResponseFormat.optional()),
    reasoningEffort: core.serialization.property("reasoning_effort", ReasoningEffort_1.ReasoningEffort.optional()),
    tools: core.serialization.list(ToolFunction_1.ToolFunction).optional(),
    linkedTools: core.serialization.property("linked_tools", core.serialization.list(core.serialization.string()).optional()),
    attributes: core.serialization.record(core.serialization.string(), core.serialization.unknown()).optional(),
});
