// humanloop/dist/decorators/prompt.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.promptDecoratorFactory = promptDecoratorFactory;
const context_1 = require("../context");
function promptDecoratorFactory(path, callable) {
    const fileType = "prompt";
    const wrappedFunction = (inputs) => {
        return context_1.HL_CONTEXT.with((0, context_1.setDecoratorContext)({
            path: path,
            type: fileType,
            version: {
                // TODO: Implement reverse lookup of template
                template: undefined,
            },
        }), () => __awaiter(this, void 0, void 0, function* () {
            return yield callable(inputs);
        }));
    };
    return Object.assign(wrappedFunction, {
        file: {
            path: path,
            type: fileType,
            version: {
                // TODO: Implement reverse lookup of template
                template: undefined,
            },
            callable: wrappedFunction,
        },
    });
}
