// humanloop/dist/serialization/types/FlowResponse.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.FlowResponse = void 0;
const serializers = __importStar(require("../index"));
const core = __importStar(require("../../core"));
const EnvironmentResponse_1 = require("./EnvironmentResponse");
const UserResponse_1 = require("./UserResponse");
const EvaluatorAggregate_1 = require("./EvaluatorAggregate");
exports.FlowResponse = core.serialization.object({
    path: core.serialization.string(),
    id: core.serialization.string(),
    directoryId: core.serialization.property("directory_id", core.serialization.string().optional()),
    attributes: core.serialization.record(core.serialization.string(), core.serialization.unknown()),
    versionName: core.serialization.property("version_name", core.serialization.string().optional()),
    versionDescription: core.serialization.property("version_description", core.serialization.string().optional()),
    name: core.serialization.string(),
    description: core.serialization.string().optional(),
    readme: core.serialization.string().optional(),
    tags: core.serialization.list(core.serialization.string()).optional(),
    versionId: core.serialization.property("version_id", core.serialization.string()),
    type: core.serialization.stringLiteral("flow").optional(),
    environments: core.serialization.list(EnvironmentResponse_1.EnvironmentResponse).optional(),
    createdAt: core.serialization.property("created_at", core.serialization.date()),
    updatedAt: core.serialization.property("updated_at", core.serialization.date()),
    createdBy: core.serialization.property("created_by", UserResponse_1.UserResponse.optional()),
    lastUsedAt: core.serialization.property("last_used_at", core.serialization.date()),
    versionLogsCount: core.serialization.property("version_logs_count", core.serialization.number()),
    evaluatorAggregates: core.serialization.property("evaluator_aggregates", core.serialization.list(EvaluatorAggregate_1.EvaluatorAggregate).optional()),
    evaluators: core.serialization
        .list(core.serialization.lazyObject(() => serializers.MonitoringEvaluatorResponse))
        .optional(),
});
