// humanloop/dist/serialization/types/FlowLogResponse.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.FlowLogResponse = void 0;
const serializers = __importStar(require("../index"));
const core = __importStar(require("../../core"));
const ChatMessage_1 = require("./ChatMessage");
const LogStatus_1 = require("./LogStatus");
exports.FlowLogResponse = core.serialization.object({
    messages: core.serialization.list(ChatMessage_1.ChatMessage).optional(),
    outputMessage: core.serialization.property("output_message", ChatMessage_1.ChatMessage.optional()),
    startTime: core.serialization.property("start_time", core.serialization.date().optional()),
    endTime: core.serialization.property("end_time", core.serialization.date().optional()),
    output: core.serialization.string().optional(),
    createdAt: core.serialization.property("created_at", core.serialization.date().optional()),
    error: core.serialization.string().optional(),
    providerLatency: core.serialization.property("provider_latency", core.serialization.number().optional()),
    stdout: core.serialization.string().optional(),
    providerRequest: core.serialization.property("provider_request", core.serialization.record(core.serialization.string(), core.serialization.unknown()).optional()),
    providerResponse: core.serialization.property("provider_response", core.serialization.record(core.serialization.string(), core.serialization.unknown()).optional()),
    inputs: core.serialization.record(core.serialization.string(), core.serialization.unknown()).optional(),
    source: core.serialization.string().optional(),
    metadata: core.serialization.record(core.serialization.string(), core.serialization.unknown()).optional(),
    logStatus: core.serialization.property("log_status", LogStatus_1.LogStatus.optional()),
    sourceDatapointId: core.serialization.property("source_datapoint_id", core.serialization.string().optional()),
    traceParentId: core.serialization.property("trace_parent_id", core.serialization.string().optional()),
    batches: core.serialization.list(core.serialization.string()).optional(),
    user: core.serialization.string().optional(),
    environment: core.serialization.string().optional(),
    save: core.serialization.boolean().optional(),
    logId: core.serialization.property("log_id", core.serialization.string().optional()),
    id: core.serialization.string(),
    evaluatorLogs: core.serialization.property("evaluator_logs", core.serialization.list(core.serialization.lazyObject(() => serializers.EvaluatorLogResponse))),
    traceFlowId: core.serialization.property("trace_flow_id", core.serialization.string().optional()),
    traceId: core.serialization.property("trace_id", core.serialization.string().optional()),
    traceChildren: core.serialization.property("trace_children", core.serialization.list(core.serialization.lazy(() => serializers.LogResponse)).optional()),
    flow: core.serialization.lazyObject(() => serializers.FlowResponse),
});
