// humanloop/dist/serialization/types/PromptCallResponse.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.PromptCallResponse = void 0;
const serializers = __importStar(require("../index"));
const core = __importStar(require("../../core"));
const ChatMessage_1 = require("./ChatMessage");
const PromptCallResponseToolChoice_1 = require("./PromptCallResponseToolChoice");
const LogStatus_1 = require("./LogStatus");
const PromptCallLogResponse_1 = require("./PromptCallLogResponse");
exports.PromptCallResponse = core.serialization.object({
    startTime: core.serialization.property("start_time", core.serialization.date().optional()),
    endTime: core.serialization.property("end_time", core.serialization.date().optional()),
    messages: core.serialization.list(ChatMessage_1.ChatMessage).optional(),
    toolChoice: core.serialization.property("tool_choice", PromptCallResponseToolChoice_1.PromptCallResponseToolChoice.optional()),
    prompt: core.serialization.lazyObject(() => serializers.PromptResponse),
    inputs: core.serialization.record(core.serialization.string(), core.serialization.unknown()).optional(),
    source: core.serialization.string().optional(),
    metadata: core.serialization.record(core.serialization.string(), core.serialization.unknown()).optional(),
    logStatus: core.serialization.property("log_status", LogStatus_1.LogStatus.optional()),
    sourceDatapointId: core.serialization.property("source_datapoint_id", core.serialization.string().optional()),
    traceParentId: core.serialization.property("trace_parent_id", core.serialization.string().optional()),
    user: core.serialization.string().optional(),
    environment: core.serialization.string().optional(),
    save: core.serialization.boolean().optional(),
    logId: core.serialization.property("log_id", core.serialization.string().optional()),
    id: core.serialization.string(),
    traceId: core.serialization.property("trace_id", core.serialization.string().optional()),
    logs: core.serialization.list(PromptCallLogResponse_1.PromptCallLogResponse),
});
