// humanloop/otel/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HUMANLOOP_CONTEXT_DECORATOR = exports.HUMANLOOP_CONTEXT_PROMPT = exports.HUMANLOOP_CONTEXT_EVALUATION = exports.HUMANLOOP_CONTEXT_TRACE_ID = exports.HUMANLOOP_TOOL_SPAN_NAME = exports.HUMANLOOP_PROMPT_SPAN_NAME = exports.HUMANLOOP_FLOW_SPAN_NAME = exports.HUMANLOOP_SPAN_PREFIX = exports.HUMANLOOP_FLOW_PREREQUISITES_KEY = exports.HUMANLOOP_PATH_KEY = exports.HUMANLOOP_FILE_TYPE_KEY = exports.HUMANLOOP_LOG_KEY = exports.HUMANLOOP_FILE_KEY = void 0;
// Attribute name prefix on Humanloop spans for file-related attributes + path
exports.HUMANLOOP_FILE_KEY = "humanloop.file";
// Attribute name prefix on Humanloop spans for log-related attributes
exports.HUMANLOOP_LOG_KEY = "humanloop.log";
exports.HUMANLOOP_FILE_TYPE_KEY = "humanloop.file.type";
exports.HUMANLOOP_PATH_KEY = "humanloop.file.path";
exports.HUMANLOOP_FLOW_PREREQUISITES_KEY = "humanloop.flow.prerequisites";
exports.HUMANLOOP_SPAN_PREFIX = "humanloop.";
exports.HUMANLOOP_FLOW_SPAN_NAME = `${exports.HUMANLOOP_SPAN_PREFIX}flow`;
exports.HUMANLOOP_PROMPT_SPAN_NAME = `${exports.HUMANLOOP_SPAN_PREFIX}prompt`;
exports.HUMANLOOP_TOOL_SPAN_NAME = `${exports.HUMANLOOP_SPAN_PREFIX}tool`;
exports.HUMANLOOP_CONTEXT_TRACE_ID = "humanloop.context.flow.trace_id";
exports.HUMANLOOP_CONTEXT_EVALUATION = "humanloop.context.evaluation";
exports.HUMANLOOP_CONTEXT_PROMPT = "humanloop.context.prompt";
exports.HUMANLOOP_CONTEXT_DECORATOR = "humanloop.context.decorator";
