// humanloop/otel/proto/trace.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SpanFlags = exports.StatusCode = exports.SpanKind = void 0;
var SpanKind;
(function (SpanKind) {
    SpanKind[SpanKind["INTERNAL"] = 0] = "INTERNAL";
    SpanKind[SpanKind["SERVER"] = 1] = "SERVER";
    SpanKind[SpanKind["CLIENT"] = 2] = "CLIENT";
    SpanKind[SpanKind["PRODUCER"] = 3] = "PRODUCER";
    SpanKind[SpanKind["CONSUMER"] = 4] = "CONSUMER";
})(SpanKind || (exports.SpanKind = SpanKind = {}));
var StatusCode;
(function (StatusCode) {
    StatusCode[StatusCode["UNSET"] = 0] = "UNSET";
    StatusCode[StatusCode["OK"] = 1] = "OK";
    StatusCode[StatusCode["ERROR"] = 2] = "ERROR";
})(StatusCode || (exports.StatusCode = StatusCode = {}));
var SpanFlags;
(function (SpanFlags) {
    SpanFlags[SpanFlags["DO_NOT_USE"] = 0] = "DO_NOT_USE";
    SpanFlags[SpanFlags["TRACE_FLAGS_MASK"] = 255] = "TRACE_FLAGS_MASK";
    SpanFlags[SpanFlags["CONTEXT_HAS_IS_REMOTE_MASK"] = 256] = "CONTEXT_HAS_IS_REMOTE_MASK";
    SpanFlags[SpanFlags["CONTEXT_IS_REMOTE_MASK"] = 512] = "CONTEXT_IS_REMOTE_MASK";
})(SpanFlags || (exports.SpanFlags = SpanFlags = {}));
