// humanloop/Client.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.HumanloopClient = void 0;
const Client_1 = require("./api/resources/prompts/client/Client");
const Client_2 = require("./api/resources/tools/client/Client");
const Client_3 = require("./api/resources/datasets/client/Client");
const Client_4 = require("./api/resources/evaluators/client/Client");
const Client_5 = require("./api/resources/flows/client/Client");
const Client_6 = require("./api/resources/directories/client/Client");
const Client_7 = require("./api/resources/files/client/Client");
const Client_8 = require("./api/resources/evaluations/client/Client");
const Client_9 = require("./api/resources/logs/client/Client");
class HumanloopClient {
    constructor(_options = {}) {
        this._options = _options;
    }
    get prompts() {
        var _a;
        return ((_a = this._prompts) !== null && _a !== void 0 ? _a : (this._prompts = new Client_1.Prompts(this._options)));
    }
    get tools() {
        var _a;
        return ((_a = this._tools) !== null && _a !== void 0 ? _a : (this._tools = new Client_2.Tools(this._options)));
    }
    get datasets() {
        var _a;
        return ((_a = this._datasets) !== null && _a !== void 0 ? _a : (this._datasets = new Client_3.Datasets(this._options)));
    }
    get evaluators() {
        var _a;
        return ((_a = this._evaluators) !== null && _a !== void 0 ? _a : (this._evaluators = new Client_4.Evaluators(this._options)));
    }
    get flows() {
        var _a;
        return ((_a = this._flows) !== null && _a !== void 0 ? _a : (this._flows = new Client_5.Flows(this._options)));
    }
    get directories() {
        var _a;
        return ((_a = this._directories) !== null && _a !== void 0 ? _a : (this._directories = new Client_6.Directories(this._options)));
    }
    get files() {
        var _a;
        return ((_a = this._files) !== null && _a !== void 0 ? _a : (this._files = new Client_7.Files(this._options)));
    }
    get evaluations() {
        var _a;
        return ((_a = this._evaluations) !== null && _a !== void 0 ? _a : (this._evaluations = new Client_8.Evaluations(this._options)));
    }
    get logs() {
        var _a;
        return ((_a = this._logs) !== null && _a !== void 0 ? _a : (this._logs = new Client_9.Logs(this._options)));
    }
}
exports.HumanloopClient = HumanloopClient;
