// humanloop/context.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.EvaluationContext = exports.HL_CONTEXT = void 0;
exports.getTraceId = getTraceId;
exports.setTraceId = setTraceId;
exports.setDecoratorContext = setDecoratorContext;
exports.getDecoratorContext = getDecoratorContext;
exports.setEvaluationContext = setEvaluationContext;
exports.getEvaluationContext = getEvaluationContext;
const otel_api = __importStar(require("@opentelemetry/api"));
const context_async_hooks_1 = require("@opentelemetry/context-async-hooks");
const error_1 = require("./error");
const constants_1 = require("./otel/constants");
exports.HL_CONTEXT = new context_async_hooks_1.AsyncHooksContextManager();
exports.HL_CONTEXT.enable();
function getTraceId() {
    const key = otel_api.createContextKey(constants_1.HUMANLOOP_CONTEXT_TRACE_ID);
    const value = exports.HL_CONTEXT.active().getValue(key);
    return (value || undefined);
}
function setTraceId(flowLogId) {
    const key = otel_api.createContextKey(constants_1.HUMANLOOP_CONTEXT_TRACE_ID);
    return exports.HL_CONTEXT.active().setValue(key, flowLogId);
}
function setDecoratorContext(decoratorContext) {
    const key = otel_api.createContextKey(constants_1.HUMANLOOP_CONTEXT_DECORATOR);
    return exports.HL_CONTEXT.active().setValue(key, decoratorContext);
}
function getDecoratorContext() {
    const key = otel_api.createContextKey(constants_1.HUMANLOOP_CONTEXT_DECORATOR);
    return (exports.HL_CONTEXT.active().getValue(key) || undefined);
}
class EvaluationContext {
    constructor({ sourceDatapointId, runId, evalCallback, fileId, path, }) {
        this.sourceDatapointId = sourceDatapointId;
        this.runId = runId;
        this._callback = evalCallback;
        this.fileId = fileId;
        this.path = path;
        this._logged = false;
    }
    get logged() {
        return this._logged;
    }
    logArgsWithContext({ logArgs, forOtel, path, fileId, }) {
        if (path === undefined && fileId === undefined) {
            throw new error_1.HumanloopRuntimeError("Internal error: Evaluation context called without providing a path or file_id");
        }
        if (this._logged) {
            return [logArgs, null];
        }
        if (this.path !== undefined && this.path === path) {
            this._logged = true;
            return [
                forOtel
                    ? Object.assign(Object.assign({}, logArgs), { source_datapoint_id: this.sourceDatapointId, run_id: this.runId }) : Object.assign(Object.assign({}, logArgs), { sourceDatapointId: this.sourceDatapointId, runId: this.runId }),
                this._callback,
            ];
        }
        else if (this.fileId !== undefined && this.fileId === fileId) {
            this._logged = true;
            return [
                forOtel
                    ? Object.assign(Object.assign({}, logArgs), { sourceDatapointId: this.sourceDatapointId, runId: this.runId }) : Object.assign(Object.assign({}, logArgs), { sourceDatapointId: this.sourceDatapointId, runId: this.runId }),
                this._callback,
            ];
        }
        else {
            return [logArgs, null];
        }
    }
}
exports.EvaluationContext = EvaluationContext;
// ... existing code ...
function setEvaluationContext(evaluationContext) {
    const key = otel_api.createContextKey(constants_1.HUMANLOOP_CONTEXT_EVALUATION);
    return exports.HL_CONTEXT.active().setValue(key, evaluationContext);
}
function getEvaluationContext() {
    const key = otel_api.createContextKey(constants_1.HUMANLOOP_CONTEXT_EVALUATION);
    return (exports.HL_CONTEXT.active().getValue(key) || undefined);
}
